File: unity/Assets/Scripts/Core/OrbContainer.cs
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Events;

namespace OrbSystem
{
    /// <summary>
    /// Runtime component attached to the player. Manages the two orb slots
    /// (Left/Right), the player's alchemic element pool, their pack, and their level.
    ///
    /// Exposes the same three core actions as the desktop prototype:
    ///   - AddMaterialToOrb()      : pick up a material into a free orb
    ///   - TryTransmute()          : attempt to combine the two orb materials
    ///   - DecomposeMaterialAt()   : alchemically break a material into elements
    ///
    /// plus what a shipping mobile loop needs on top: a pack to hold results,
    /// loading orbs back out of that pack, and XP.
    ///
    /// This script intentionally has no UI dependency. UI listens to the
    /// UnityEvents and queries the public properties; the orb logic stays clean.
    /// </summary>
    public class OrbContainer : MonoBehaviour
    {
        public enum Hand { Left, Right }

        [System.Serializable] public class MaterialEvent : UnityEvent<MaterialSO> { }
        [System.Serializable] public class RecipeEvent : UnityEvent<TransmutationRecipe> { }
        [System.Serializable] public class AlchemyEvent : UnityEvent<AlchemyRecipe> { }
        [System.Serializable] public class NoticeEvent : UnityEvent<string> { }

        [Header("Content")]
        [Tooltip("Loads elements, materials and recipes from the shared content JSON. " +
                 "Leave empty to use hand-authored assets in Resources/ instead.")]
        public ContentDatabase contentDatabase;

        [Header("Player State")]
        [Tooltip("Level at which alchemic decomposition unlocks.")]
        public int alchemyUnlockLevel = 5;

        [Header("Spawn Settings")]
        [Tooltip("Where transmuted/alchemized objects spawn (e.g., in front of the player).")]
        public Transform spawnPoint;

        [Tooltip("Also Instantiate() a recipe's resultPrefab, when it has one. " +
                 "Off by default: on mobile the result normally just goes into the pack.")]
        public bool spawnResultPrefabs;

        // ---- Runtime State ----
        private MaterialSO leftOrb;
        private MaterialSO rightOrb;
        private readonly Dictionary<ElementSO, int> elementPool = new Dictionary<ElementSO, int>();

        public readonly PlayerProgression progression = new PlayerProgression();
        public readonly Inventory pack = new Inventory();

        private readonly HashSet<string> discoveredRecipes = new HashSet<string>();
        private readonly HashSet<string> seenMaterials = new HashSet<string>();

        // ---- Events for UI to subscribe to ----
        public UnityEvent OnOrbContentsChanged = new UnityEvent();
        public UnityEvent OnElementPoolChanged = new UnityEvent();
        public MaterialEvent OnMaterialGathered = new MaterialEvent();
        public MaterialEvent OnMaterialCrafted = new MaterialEvent();
        public MaterialEvent OnMaterialDiscovered = new MaterialEvent();
        public MaterialEvent OnMaterialDecomposed = new MaterialEvent();
        public NoticeEvent OnNotice = new NoticeEvent();

        // ---- Public Read-Only Accessors ----
        public MaterialSO LeftOrb => leftOrb;
        public MaterialSO RightOrb => rightOrb;
        public IReadOnlyDictionary<ElementSO, int> ElementPool => elementPool;
        public int PlayerLevel => progression.Level;
        public bool AlchemyUnlocked => progression.Level >= alchemyUnlockLevel;
        public IReadOnlyCollection<string> DiscoveredRecipes => discoveredRecipes;
        public IReadOnlyCollection<string> SeenMaterials => seenMaterials;

        private void Awake()
        {
            if (contentDatabase != null)
            {
                contentDatabase.Load();
                alchemyUnlockLevel = contentDatabase.Progression.alchemyUnlockLevel;
                progression.Configure(contentDatabase.Progression.xpTable, contentDatabase.Progression.maxLevel);
            }
            else
            {
                // No JSON database wired up: fall back to hand-authored assets,
                // exactly as the desktop prototype did.
                TransmutationSystem.Initialize();
                AlchemySystem.Initialize();
            }
        }

        // ---------------------------------------------------------------
        // Adding materials
        // ---------------------------------------------------------------

        /// <summary>
        /// Place a material into the specified orb.
        /// Returns true if successful, false if the orb is already full.
        /// </summary>
        public bool AddMaterialToOrb(Hand hand, MaterialSO material)
        {
            if (material == null) return false;

            if (hand == Hand.Left)
            {
                if (leftOrb != null) return false;
                leftOrb = material;
            }
            else
            {
                if (rightOrb != null) return false;
                rightOrb = material;
            }

            OnOrbContentsChanged.Invoke();
            return true;
        }

        /// <summary>
        /// Auto-place into whichever orb is empty. Returns the hand used,
        /// or null if both orbs are full.
        /// </summary>
        public Hand? AddMaterialToFreeOrb(MaterialSO material)
        {
            if (leftOrb == null && AddMaterialToOrb(Hand.Left, material)) return Hand.Left;
            if (rightOrb == null && AddMaterialToOrb(Hand.Right, material)) return Hand.Right;
            return null;
        }

        /// <summary>Discards the orb's contents outright.</summary>
        public void ClearOrb(Hand hand)
        {
            if (hand == Hand.Left) leftOrb = null;
            else rightOrb = null;

            OnOrbContentsChanged.Invoke();
        }

        /// <summary>Move an orb's material into the pack rather than losing it.</summary>
        public bool UnloadOrb(Hand hand)
        {
            MaterialSO material = GetOrb(hand);
            if (material == null) return false;

            SetOrb(hand, null);
            pack.Add(material);
            OnOrbContentsChanged.Invoke();
            return true;
        }

        /// <summary>Draw from the pack into an orb - how tier-2+ crafting is actually done.</summary>
        public bool LoadFromPack(Hand hand, MaterialSO material)
        {
            if (material == null) return false;
            if (GetOrb(hand) != null) return false;
            if (!pack.Remove(material)) return false;

            SetOrb(hand, material);
            OnOrbContentsChanged.Invoke();
            return true;
        }

        public Hand? LoadFromPackToFreeOrb(MaterialSO material)
        {
            if (leftOrb == null && LoadFromPack(Hand.Left, material)) return Hand.Left;
            if (rightOrb == null && LoadFromPack(Hand.Right, material)) return Hand.Right;
            return null;
        }

        public void SwapOrbs()
        {
            (leftOrb, rightOrb) = (rightOrb, leftOrb);
            OnOrbContentsChanged.Invoke();
        }

        // ---------------------------------------------------------------
        // Gathering
        // ---------------------------------------------------------------

        /// <summary>
        /// Pick a material up out of the world. Falls through to the pack when
        /// both orbs are full, so a pickup is never silently wasted.
        /// </summary>
        public Hand? Gather(MaterialSO material)
        {
            if (material == null) return null;

            bool isNew = seenMaterials.Add(material.materialID);
            Hand? hand = AddMaterialToFreeOrb(material);
            if (hand == null) pack.Add(material);

            var xpConfig = contentDatabase != null ? contentDatabase.Progression.xp : null;
            int gain = xpConfig == null ? 4 : (isNew ? xpConfig.gatherNewMaterial : xpConfig.gather);
            progression.AddXp(gain);

            OnMaterialGathered.Invoke(material);
            if (isNew) OnMaterialDiscovered.Invoke(material);
            else if (hand == null) OnNotice.Invoke($"Orbs full - {material.displayName} went to your pack");

            return hand;
        }

        // ---------------------------------------------------------------
        // Transmutation
        // ---------------------------------------------------------------

        /// <summary>
        /// Peek at what would result from transmuting the current orb contents.
        /// Useful for showing a preview in the UI before committing.
        /// Returns null if no valid recipe exists.
        /// </summary>
        public TransmutationRecipe PeekTransmutation()
        {
            if (leftOrb == null || rightOrb == null) return null;
            return TransmutationSystem.FindRecipe(leftOrb, rightOrb, progression.Level);
        }

        /// <summary>
        /// A recipe exists for the current pair but the player is too low a level.
        /// Lets the UI say "needs level 12" instead of "no reaction".
        /// </summary>
        public TransmutationRecipe PeekLockedTransmutation()
        {
            if (leftOrb == null || rightOrb == null) return null;
            return TransmutationSystem.FindLockedRecipe(leftOrb, rightOrb, progression.Level);
        }

        /// <summary>
        /// Commit to the transmutation: consumes both orb materials and grants the
        /// result. Returns the resulting material, or null if no valid recipe.
        /// </summary>
        public MaterialSO TryTransmute()
        {
            var recipe = PeekTransmutation();
            if (recipe == null || recipe.resultMaterial == null) return null;

            // Consume inputs
            leftOrb = null;
            rightOrb = null;
            OnOrbContentsChanged.Invoke();

            bool isNew = discoveredRecipes.Add(recipe.recipeID);
            GrantResult(recipe.resultMaterial, recipe.resultPrefab);

            float factor = 1f;
            if (!isNew && contentDatabase != null) factor = contentDatabase.Progression.xp.repeatTransmuteFactor;
            progression.AddXp(Mathf.Max(1, Mathf.RoundToInt(recipe.xp * factor)));

            if (isNew) OnMaterialDiscovered.Invoke(recipe.resultMaterial);
            OnMaterialCrafted.Invoke(recipe.resultMaterial);
            return recipe.resultMaterial;
        }

        // ---------------------------------------------------------------
        // Alchemy
        // ---------------------------------------------------------------

        /// <summary>
        /// Alchemically decompose the material in the specified orb into its
        /// constituent elements, which are added to the player's element pool.
        /// Locked behind alchemyUnlockLevel.
        /// </summary>
        public bool DecomposeMaterialAt(Hand hand)
        {
            if (!AlchemyUnlocked)
            {
                OnNotice.Invoke($"Alchemy unlocks at level {alchemyUnlockLevel}");
                return false;
            }

            MaterialSO target = GetOrb(hand);
            if (target == null) return false;

            foreach (var comp in target.elementComposition)
            {
                if (comp.element == null || comp.quantity <= 0) continue;

                elementPool.TryGetValue(comp.element, out int have);
                elementPool[comp.element] = have + comp.quantity;
            }

            ClearOrb(hand);

            int gain = contentDatabase != null ? contentDatabase.Progression.xp.decompose : 6;
            progression.AddXp(gain);

            OnElementPoolChanged.Invoke();
            OnMaterialDecomposed.Invoke(target);
            return true;
        }

        /// <summary>
        /// Get every alchemy recipe the player can currently afford with their
        /// element pool. Use this to populate the alchemy menu.
        /// </summary>
        public List<AlchemyRecipe> GetAvailableAlchemyRecipes()
        {
            if (!AlchemyUnlocked) return new List<AlchemyRecipe>();
            return AlchemySystem.FindAvailableRecipes(elementPool, progression.Level);
        }

        public bool CanAfford(AlchemyRecipe recipe)
        {
            return AlchemySystem.CanFulfill(recipe, elementPool);
        }

        public Dictionary<ElementSO, int> Shortfall(AlchemyRecipe recipe)
        {
            return AlchemySystem.Shortfall(recipe, elementPool);
        }

        /// <summary>
        /// Commit to an alchemy recipe: consumes the required elements and grants
        /// the result. Returns the resulting material, or null on failure.
        /// </summary>
        public MaterialSO TryAlchemize(AlchemyRecipe recipe)
        {
            if (recipe == null || !AlchemyUnlocked) return null;
            if (recipe.requiredLevel > progression.Level) return null;
            if (recipe.resultMaterial == null) return null;
            if (!AlchemySystem.ConsumeElements(recipe, elementPool)) return null;

            bool isNew = discoveredRecipes.Add(recipe.recipeID);
            GrantResult(recipe.resultMaterial, recipe.resultPrefab);

            float factor = 1f;
            if (!isNew && contentDatabase != null) factor = contentDatabase.Progression.xp.repeatAlchemyFactor;
            progression.AddXp(Mathf.Max(1, Mathf.RoundToInt(recipe.xp * factor)));

            OnElementPoolChanged.Invoke();
            if (isNew) OnMaterialDiscovered.Invoke(recipe.resultMaterial);
            OnMaterialCrafted.Invoke(recipe.resultMaterial);
            return recipe.resultMaterial;
        }

        // ---------------------------------------------------------------
        // Helpers
        // ---------------------------------------------------------------

        public MaterialSO GetOrb(Hand hand)
        {
            return hand == Hand.Left ? leftOrb : rightOrb;
        }

        private void SetOrb(Hand hand, MaterialSO material)
        {
            if (hand == Hand.Left) leftOrb = material;
            else rightOrb = material;
        }

        public bool HasSeen(MaterialSO material)
        {
            return material != null && seenMaterials.Contains(material.materialID);
        }

        public bool HasDiscovered(string recipeID)
        {
            return discoveredRecipes.Contains(recipeID);
        }

        private void GrantResult(MaterialSO material, GameObject prefab)
        {
            pack.Add(material);
            seenMaterials.Add(material.materialID);

            if (!spawnResultPrefabs || prefab == null) return;

            Transform t = spawnPoint != null ? spawnPoint : transform;
            Instantiate(prefab, t.position, t.rotation);
        }
    }
}

File: web/src/core/orbContainer.ts
/**
 * Port of OrbContainer.cs - the two orb slots, the element pool, and the three
 * core actions (add / transmute / decompose), plus the state the mobile build
 * needs that the Unity prototype left to the scene: an inventory, XP, and the
 * discovery log.
 *
 * As in the original this file has no UI dependency. It emits events and exposes
 * plain accessors; the renderer and HUD subscribe. Deliberate differences from
 * the C# version, all of them forced by making this an actual game rather than a
 * system sketch:
 *
 *   - Results are material ids added to an inventory, not Instantiate()d prefabs.
 *     Crafted things have to be able to go back into an orb, or the tree can't chain.
 *   - Orbs can be loaded from the inventory, not just from world pickups.
 *   - Decomposition is lossy (progression.decompositionYield), so decompose ->
 *     recompose can't be farmed for free elements.
 */
import { Emitter } from './events';
import type { Content } from './content';
import {
  canFulfill,
  consumeElements,
  findAvailableRecipes,
  findRecipeForSelection,
  type ElementPool,
} from './alchemy';
import { findRecipe } from './transmutation';
import { levelForXp } from './progression';
import type {
  AlchemyRecipe,
  ElementId,
  Hand,
  MaterialId,
  RecipeId,
  TransmutationRecipe,
  ZoneId,
} from './types';

export interface GameStats {
  gathered: number;
  transmuted: number;
  alchemized: number;
  decomposed: number;
}

export interface GameState {
  level: number;
  xp: number;
  zoneId: ZoneId;
  orbs: Record<Hand, MaterialId | null>;
  elementPool: ElementPool;
  inventory: Record<MaterialId, number>;
  discovered: Set<RecipeId>;
  seenMaterials: Set<MaterialId>;
  stats: GameStats;
  playtimeMs: number;
}

export type XpReason = 'gather' | 'discover' | 'transmute' | 'alchemy' | 'decompose';

export interface OrbEvents {
  orbsChanged: void;
  poolChanged: void;
  inventoryChanged: void;
  stateChanged: void;
  xpGained: { amount: number; reason: XpReason };
  levelUp: { level: number; unlockedAlchemy: boolean; unlockedZones: ZoneId[] };
  discovered: { recipe: TransmutationRecipe | AlchemyRecipe; material: MaterialId };
  crafted: { material: MaterialId; via: 'transmutation' | 'alchemy'; isNew: boolean };
  gathered: { material: MaterialId; isNew: boolean; toOrb: Hand | null };
  decomposed: { material: MaterialId; gained: Record<ElementId, number> };
  notice: { text: string; tone: 'info' | 'good' | 'bad' };
  zoneChanged: { zoneId: ZoneId };
}

export class OrbContainer {
  readonly events = new Emitter<OrbEvents>();

  constructor(
    private readonly content: Content,
    readonly state: GameState,
  ) {}

  // ---------------------------------------------------------------- accessors

  get leftOrb(): MaterialId | null {
    return this.state.orbs.left;
  }

  get rightOrb(): MaterialId | null {
    return this.state.orbs.right;
  }

  get playerLevel(): number {
    return this.state.level;
  }

  /** OrbContainer.AlchemyUnlocked */
  get alchemyUnlocked(): boolean {
    return this.state.level >= this.content.progression.alchemyUnlockLevel;
  }

  orb(hand: Hand): MaterialId | null {
    return this.state.orbs[hand];
  }

  countOf(material: MaterialId): number {
    return this.state.inventory[material] ?? 0;
  }

  /** Inventory contents, deepest-tier first so late-game items surface at the top. */
  packContents(): { material: MaterialId; count: number }[] {
    return Object.entries(this.state.inventory)
      .filter(([, count]) => count > 0)
      .map(([material, count]) => ({ material, count }))
      .sort((x, y) => {
        const mx = this.content.material(x.material);
        const my = this.content.material(y.material);
        return my.tier - mx.tier || mx.name.localeCompare(my.name);
      });
  }

  // ---------------------------------------------------------------- orbs

  /** OrbContainer.AddMaterialToOrb - fails if that orb is already full. */
  addMaterialToOrb(hand: Hand, material: MaterialId): boolean {
    if (!this.content.hasMaterial(material)) return false;
    if (this.state.orbs[hand] !== null) return false;

    this.state.orbs[hand] = material;
    this.events.emit('orbsChanged', undefined);
    return true;
  }

  /** OrbContainer.AddMaterialToFreeOrb - returns the hand used, or null if both are full. */
  addMaterialToFreeOrb(material: MaterialId): Hand | null {
    if (this.addMaterialToOrb('left', material)) return 'left';
    if (this.addMaterialToOrb('right', material)) return 'right';
    return null;
  }

  /** OrbContainer.ClearOrb - discards the contents outright. */
  clearOrb(hand: Hand): void {
    this.state.orbs[hand] = null;
    this.events.emit('orbsChanged', undefined);
  }

  /** Move an orb's material back into the pack rather than losing it. */
  unloadOrb(hand: Hand): boolean {
    const material = this.state.orbs[hand];
    if (material === null) return false;

    this.state.orbs[hand] = null;
    this.addToInventory(material, 1);
    this.events.emit('orbsChanged', undefined);
    return true;
  }

  /** Draw from the pack into an orb - how tier-2+ crafting is actually done. */
  loadFromInventory(hand: Hand, material: MaterialId): boolean {
    if (this.countOf(material) <= 0) return false;
    if (this.state.orbs[hand] !== null) return false;

    this.removeFromInventory(material, 1);
    this.state.orbs[hand] = material;
    this.events.emit('orbsChanged', undefined);
    return true;
  }

  /** Load into whichever orb is free. */
  loadFromInventoryToFreeOrb(material: MaterialId): Hand | null {
    if (this.loadFromInventory('left', material)) return 'left';
    if (this.loadFromInventory('right', material)) return 'right';
    return null;
  }

  /**
   * Put a material into an orb whether or not it already holds something,
   * returning whatever was there to the pack. This is what lets the bench swap
   * one input for another without a clear-then-load dance.
   */
  replaceOrb(hand: Hand, material: MaterialId): boolean {
    if (this.countOf(material) <= 0) return false;

    const current = this.state.orbs[hand];
    if (current === material) return false;

    this.removeFromInventory(material, 1);
    if (current !== null) this.addToInventory(current, 1);

    this.state.orbs[hand] = material;
    this.events.emit('orbsChanged', undefined);
    this.events.emit('stateChanged', undefined);
    return true;
  }

  swapOrbs(): void {
    const { left, right } = this.state.orbs;
    this.state.orbs.left = right;
    this.state.orbs.right = left;
    this.events.emit('orbsChanged', undefined);
  }

  // ---------------------------------------------------------------- gathering

  /** Pick a material up out of the world. Falls through to the pack when both orbs are full. */
  gather(material: MaterialId): { toOrb: Hand | null; isNew: boolean } {
    const isNew = !this.state.seenMaterials.has(material);
    const toOrb = this.addMaterialToFreeOrb(material);
    if (toOrb === null) this.addToInventory(material, 1);

    this.state.stats.gathered++;
    this.markSeen(material);

    const xpConfig = this.content.progression.xp;
    this.addXp(isNew ? xpConfig.gatherNewMaterial : xpConfig.gather, isNew ? 'discover' : 'gather');

    this.events.emit('gathered', { material, isNew, toOrb });
    this.events.emit('stateChanged', undefined);
    return { toOrb, isNew };
  }

  // ---------------------------------------------------------------- transmutation

  /** OrbContainer.PeekTransmutation - what the current orb pair would produce. */
  peekTransmutation(): TransmutationRecipe | null {
    return findRecipe(this.content, this.leftOrb, this.rightOrb, this.state.level);
  }

  /**
   * OrbContainer.TryTransmute - consumes both orbs and produces the result.
   * Returns the resulting material id, or null when the pair has no usable recipe.
   */
  tryTransmute(): MaterialId | null {
    const recipe = this.peekTransmutation();
    if (recipe === null) return null;

    this.state.orbs.left = null;
    this.state.orbs.right = null;
    this.events.emit('orbsChanged', undefined);

    const isNew = !this.state.discovered.has(recipe.id);
    this.state.discovered.add(recipe.id);
    this.state.stats.transmuted++;

    this.addToInventory(recipe.result, 1);
    this.markSeen(recipe.result);

    const factor = isNew ? 1 : this.content.progression.xp.repeatTransmuteFactor;
    this.addXp(Math.max(1, Math.round(recipe.xp * factor)), isNew ? 'discover' : 'transmute');

    if (isNew) this.events.emit('discovered', { recipe, material: recipe.result });
    this.events.emit('crafted', { material: recipe.result, via: 'transmutation', isNew });
    this.events.emit('stateChanged', undefined);
    return recipe.result;
  }

  // ---------------------------------------------------------------- alchemy

  /**
   * OrbContainer.DecomposeMaterialAt - break an orb's material into its elements.
   * Gated behind alchemyUnlockLevel, exactly as the original.
   */
  decomposeMaterialAt(hand: Hand): boolean {
    if (!this.alchemyUnlocked) return false;

    const material = this.state.orbs[hand];
    if (material === null) return false;

    const gained: Record<ElementId, number> = {};
    for (const [element, quantity] of Object.entries(this.content.material(material).composition)) {
      if (quantity <= 0) continue;
      this.state.elementPool[element] = (this.state.elementPool[element] ?? 0) + quantity;
      gained[element] = quantity;
    }

    this.clearOrb(hand);
    this.state.stats.decomposed++;
    this.addXp(this.content.progression.xp.decompose, 'decompose');

    this.events.emit('poolChanged', undefined);
    this.events.emit('decomposed', { material, gained });
    this.events.emit('stateChanged', undefined);
    return true;
  }

  /** OrbContainer.GetAvailableAlchemyRecipes - populates the alchemy menu. */
  getAvailableAlchemyRecipes(): readonly AlchemyRecipe[] {
    if (!this.alchemyUnlocked) return [];
    return findAvailableRecipes(this.content, this.state.elementPool, this.state.level);
  }

  canAfford(recipe: AlchemyRecipe): boolean {
    return canFulfill(recipe, this.state.elementPool);
  }

  /**
   * Attempt a hand-mixed combination. Returns the result, or null when the
   * selection matches no recipe.
   *
   * A failed mix costs nothing. With nine elements and free quantities the
   * search space is large, and charging for wrong guesses would make
   * experimenting - the entire point of the table - feel punishing.
   */
  tryAlchemizeSelection(selection: Readonly<Record<ElementId, number>>): {
    result: MaterialId | null;
    reason: 'ok' | 'no-recipe' | 'locked' | 'short';
  } {
    if (!this.alchemyUnlocked) return { result: null, reason: 'locked' };

    const recipe = findRecipeForSelection(this.content, selection, this.state.level);
    if (!recipe) return { result: null, reason: 'no-recipe' };
    if (!canFulfill(recipe, this.state.elementPool)) return { result: null, reason: 'short' };

    return { result: this.tryAlchemize(recipe), reason: 'ok' };
  }

  /** OrbContainer.TryAlchemize - spends the elements and produces the result. */
  tryAlchemize(recipe: AlchemyRecipe): MaterialId | null {
    if (!this.alchemyUnlocked) return null;
    if (recipe.requiredLevel > this.state.level) return null;
    if (!consumeElements(recipe, this.state.elementPool)) return null;

    const isNew = !this.state.discovered.has(recipe.id);
    this.state.discovered.add(recipe.id);
    this.state.stats.alchemized++;

    this.addToInventory(recipe.result, 1);
    this.markSeen(recipe.result);

    const factor = isNew ? 1 : this.content.progression.xp.repeatAlchemyFactor;
    this.addXp(Math.max(1, Math.round(recipe.xp * factor)), isNew ? 'discover' : 'alchemy');

    this.events.emit('poolChanged', undefined);
    if (isNew) this.events.emit('discovered', { recipe, material: recipe.result });
    this.events.emit('crafted', { material: recipe.result, via: 'alchemy', isNew });
    this.events.emit('stateChanged', undefined);
    return recipe.result;
  }

  // ---------------------------------------------------------------- travel

  travelTo(zoneId: ZoneId): boolean {
    const zone = this.content.zone(zoneId);
    if (zone.requiredLevel > this.state.level) return false;
    if (this.state.zoneId === zoneId) return false;

    this.state.zoneId = zoneId;
    this.events.emit('zoneChanged', { zoneId });
    this.events.emit('stateChanged', undefined);
    return true;
  }

  // ---------------------------------------------------------------- progression

  addXp(amount: number, reason: XpReason): void {
    if (amount <= 0) return;

    const before = this.state.level;
    this.state.xp += amount;
    const after = levelForXp(this.content.progression, this.state.xp);
    this.events.emit('xpGained', { amount, reason });

    if (after <= before) return;

    this.state.level = after;
    const unlockLevel = this.content.progression.alchemyUnlockLevel;
    const unlockedZones = this.content.zones
      .filter((z) => z.requiredLevel > before && z.requiredLevel <= after)
      .map((z) => z.id);

    this.events.emit('levelUp', {
      level: after,
      unlockedAlchemy: before < unlockLevel && after >= unlockLevel,
      unlockedZones,
    });
  }

  // ---------------------------------------------------------------- inventory

  private addToInventory(material: MaterialId, count: number): void {
    this.state.inventory[material] = this.countOf(material) + count;
    this.events.emit('inventoryChanged', undefined);
  }

  private removeFromInventory(material: MaterialId, count: number): void {
    const remaining = this.countOf(material) - count;
    if (remaining > 0) this.state.inventory[material] = remaining;
    else delete this.state.inventory[material];
    this.events.emit('inventoryChanged', undefined);
  }

  private markSeen(material: MaterialId): void {
    this.state.seenMaterials.add(material);
  }
}

/** A fresh run: level 1, empty orbs, standing in the first zone. */
export function createInitialState(content: Content): GameState {
  const firstZone = content.zones[0];
  if (!firstZone) throw new Error('content bundle defines no zones');

  return {
    level: 1,
    xp: 0,
    zoneId: firstZone.id,
    orbs: { left: null, right: null },
    elementPool: {},
    inventory: {},
    discovered: new Set(),
    seenMaterials: new Set(),
    stats: { gathered: 0, transmuted: 0, alchemized: 0, decomposed: 0 },
    playtimeMs: 0,
  };
}
